PA.tex

```diff
@@ -0,0 +1,173 @@
+\section{Optimisation for Processor Architectures
+\label{sec:pa}}
+
+\subsection{NVIDIA GPU}
+The LFRic-microbenchmark suite~\cite{lfric-microbenchmarks} allows
+optimisations for individual kernels to be explored without having to
+compile and run the whole code. This is particularly useful when
+targeting a programming model such as OpenACC for GPU architecture
+where the Portland group compiler cannot compile the Fortran 2003 used 
+in the main code.
+
+The local matrix assembly (LMA) matrix-vector kernel had been
+previously ported to run on GPUs by inserting Open ACC directives in
+the PSy layer code to control the data transfer from host memory to
+device memory and to offload the kernel call to the GPU device. This
+Open ACC version was the starting point for Alan Gray from NVIDIA to
+to look at optimisations. Moreover, after discussions regarding 
+the PSyKAl API and the LFRic plans for how Open ACC will be
+implemented for a GPU version, NVIDIA believe that the plan and
+implementation will enable a succesful GPU implementation.
+This is a summary of potential optimisations from NVIDIA.
+
+\begin{lstlisting}[language=Fortran,caption={Code fragment of original 
+    kernel},label={lst:LMA-orig}]
+  !$acc loop vector 
+  do k = 0, nlayers-1
+    lhs_e(:,:) = 0.0_r_def
+    ik = (cell-1)*nlayers + k + 1
+
+    do df = 1, ndf1
+      do df2 = 1, ndf2
+        lhs_e(df,k) = lhs_e(df,k) + & 
+          matrix(df,df2,ik)*x(map2(df2)+k)
+      end do
+    end do
+
+    do df = 1, ndf1
+      lhs(map1(df)+k) = lhs(map1(df)+k) + lhs_e(df,k)
+    end do
+
+  end do 
+  !$
```

Comment:
As discussed, this sentinel isn't required in the final text

Reply:
Yup. Deleted.